### PR TITLE
magento/magento2#12695: Unable to change attribute type from swatch

### PR DESCRIPTION
**Pull request by eugene-shab**
Unable to change attribute type from swatch to dropdown

- fixed the process of switching attribute input type

### Description
Fixed Magento _Swatches module.

### Fixed Issues (if relevant)
1. magento/magento2#12695: Unable to change attribute type from swatch to dropdown
2. magento/magento2#11703: Changing Swatches to Drop-down does not remove swatches from existing products 
3. magento/magento2#9307: Color attribute taking swatch instead of Drop down option for configurable options
4. magento/magento2#11403: Product Attributes Not Updating on Frontend 
5. magento/magento2#9923: Upgrading to 2.1.7 changed dropdown attributes to swatches

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=12771) <br/>All committers have signed the CLA.

File: app/code/Magento/Swatches/Model/Plugin/EavAttribute.php [hunk 1]
```diff
@@ -18,6 +19,11 @@ class EavAttribute
 {
     const DEFAULT_STORE_ID = 0;
 
+    /**
+     * @var SwatchResource
+     */
+    protected $swatchResource;
```
Comment (mslabko) [inline app/code/Magento/Swatches/Model/Plugin/EavAttribute.php]: Use private

File: app/code/Magento/Swatches/Model/Plugin/EavAttribute.php [hunk 2]
```diff
@@ -68,13 +74,15 @@ class EavAttribute
     public function __construct(
         \Magento\Swatches\Model\ResourceModel\Swatch\CollectionFactory $collectionFactory,
         \Magento\Swatches\Model\SwatchFactory $swatchFactory,
+        SwatchResource $swatchResource,
```
Comment (mslabko) [inline app/code/Magento/Swatches/Model/Plugin/EavAttribute.php]: For make it backward compatible, set new dependency as optional and use ObjectManager to instantiating it

File: app/code/Magento/Swatches/Model/Plugin/EavAttribute.php [hunk 3]
```diff
@@ -157,13 +165,35 @@ protected function convertSwatchToDropdown(Attribute $attribute)
             if (!empty($additionalData)) {
                 $additionalData = $this->serializer->unserialize($additionalData);
                 if (is_array($additionalData) && isset($additionalData[Swatch::SWATCH_INPUT_TYPE_KEY])) {
+                    if ($additionalData[Swatch::SWATCH_INPUT_TYPE_KEY] == Swatch::SWATCH_INPUT_TYPE_VISUAL) {
```
Comment (mslabko) [inline app/code/Magento/Swatches/Model/Plugin/EavAttribute.php]: This issue relevant only for visusal swatches?

File: app/code/Magento/Swatches/Model/Plugin/EavAttribute.php [hunk 4]
```diff
@@ -157,13 +165,35 @@ protected function convertSwatchToDropdown(Attribute $attribute)
...
+     *
+     * @param Attribute $attribute
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    private function cleanEavAttributeOptionSwatchValues(Attribute $attribute)
+    {
+        $optionsIDs = [];
+        if (count($attribute->getOption())) {
+            $options = $attribute->getOption();
+            foreach ($options['value'] as $id => $item) {
```
Comment (mslabko) [inline app/code/Magento/Swatches/Model/Plugin/EavAttribute.php]: array_column can be used for simplify code

File: app/code/Magento/Swatches/Model/Plugin/EavAttribute.php [hunk 5]
```diff
@@ -267,6 +298,27 @@ protected function processVisualSwatch(Attribute $attribute)
...
+     *
+     * @param $attribute
+     */
+    private function cleanTextSwatchValuesAfterSwitch( Attribute $attribute)
+    {
+        $optionsIDs = [];
+        if (count($attribute->getOptiontext())) {
+            $options = $attribute->getOptiontext();
+            if (count($options)) {
+                foreach ($options['value'] as $id => $item) {
```
Comment (mslabko) [inline app/code/Magento/Swatches/Model/Plugin/EavAttribute.php]: array_column can be used for simplify code

File: app/code/Magento/Swatches/Model/Plugin/EavAttribute.php [hunk 6]
```diff
@@ -267,6 +298,27 @@ protected function processVisualSwatch(Attribute $attribute)
...
+                    $optionsIDs[] = $id;
+                }
+
+                $type = Swatch::SWATCH_TYPE_TEXTUAL;
+                $this->swatchResource->clearSwatchOptionTextByOptionId($optionsIDs, $type);
```
Comment (mslabko) [inline app/code/Magento/Swatches/Model/Plugin/EavAttribute.php]: It's better do not introduce property $type in this case - just pass constant directly to method call

File: app/code/Magento/Swatches/Model/ResourceModel/Swatch.php [hunk 7]
```diff
@@ -37,4 +37,28 @@ public function saveDefaultSwatchOption($id, $defaultValue)
...
+    {
+        if (count($optionIDs)) {
+            foreach ($optionIDs as $optionId) {
+                $where = ['option_id' => $optionId];
+                $this->getConnection()->delete($this->getMainTable(), $where);
+            }
+        }
+    }
+
+    public function clearSwatchOptionTextByOptionId($optionIDs, $type)
```
Comment (mslabko) [inline app/code/Magento/Swatches/Model/ResourceModel/Swatch.php]: Doc missed.

Can we merge both methods? It looks like "type" is redundant if we already have option id for delete

File: app/code/Magento/Swatches/Model/SwatchAttributeCodes.php [hunk 8]
```diff
@@ -67,7 +67,7 @@ public function __construct(
      *
      * @return array
      */
-    public function getCodes()
+        public function getCodes()
```
Comment (mslabko) [inline app/code/Magento/Swatches/Model/SwatchAttributeCodes.php]: please, remove extra spaces

File: app/code/Magento/Swatches/Model/SwatchAttributesProvider.php [hunk 9]
```diff
@@ -36,10 +40,12 @@ class SwatchAttributesProvider
      */
     public function __construct(
         Configurable $typeConfigurable,
-        SwatchAttributeCodes $swatchAttributeCodes
+        SwatchAttributeCodes $swatchAttributeCodes,
+        \Magento\Swatches\Helper\Data\Proxy $swatchHelper
     ) {
         $this->typeConfigurable = $typeConfigurable;
         $this->swatchAttributeCodes = $swatchAttributeCodes;
+        $this->swatchHelper = $swatchHelper;
```
Comment (mslabko) [inline app/code/Magento/Swatches/Model/SwatchAttributesProvider.php]: For make it backward compatible, set new dependency as optional and use ObjectManager to instantiating it

File: app/code/Magento/Swatches/Model/SwatchAttributesProvider.php [hunk 10]
```diff
@@ -60,9 +66,13 @@ public function provide(Product $product)
 
             $swatchAttributes = [];
             foreach ($configurableAttributes as $configurableAttribute) {
-                if (array_key_exists($configurableAttribute->getAttributeId(), $swatchAttributeCodeMap)) {
-                    $swatchAttributes[$configurableAttribute->getAttributeId()]
-                        = $configurableAttribute->getProductAttribute();
+                // Due to $this->swatchAttributeCodes->getCodes() - return swatch data for every attribute.
+                if ($this->swatchHelper->isSwatchAttribute($configurableAttribute->getProductAttribute())) {
```
Comment (mslabko) [inline app/code/Magento/Swatches/Model/SwatchAttributesProvider.php]: For more readability, these conditions can be moved to private method with description of intention of extra condition

File: app/code/Magento/Swatches/Model/SwatchAttributesProvider.php [hunk 11]
```diff
@@ -36,10 +40,12 @@ class SwatchAttributesProvider
      */
     public function __construct(
         Configurable $typeConfigurable,
-        SwatchAttributeCodes $swatchAttributeCodes
+        SwatchAttributeCodes $swatchAttributeCodes,
+        \Magento\Swatches\Helper\Data\Proxy $swatchHelper
```
Comment (mslabko) [inline app/code/Magento/Swatches/Model/SwatchAttributesProvider.php]: Please, set proxy $swatchHelper  in di.xml (remove \Proxy from construct and move it to di.xml)

Proxies and interceptors MUST NEVER be explicitly requested in constructors. See:
http://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html

**Review by mslabko — CHANGES_REQUESTED** [review]
@eugene-shab , Thank you for the contribution!

Please, check review comments

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]
Please see review by @mslabko

File: app/code/Magento/Swatches/Model/SwatchAttributesProvider.php (hunk 10 above)
Comment (omiroshnichenko) [reply app/code/Magento/Swatches/Model/SwatchAttributesProvider.php]: Also, mark this method as deprecated and add comment that this method introduced only for case when customer already has converted attribute.

File: app/code/Magento/Swatches/Model/Plugin/EavAttribute.php (hunk 3 above)
Comment (eugene-shab) [reply app/code/Magento/Swatches/Model/Plugin/EavAttribute.php]: Yes. Only when switching from visual to dropdown input type.

**Comment by sidolov** [comment]
Hi @eugene-shab , please, merge latest mainline to your branch, resolve conflicts and we continue work with PR. Thanks!

**Comment by eugene-shab** [comment]
After last changes in 2.2-develop: it seems that method hasSwatchAttribute can be removed.

Please continue.

File: app/code/Magento/Swatches/Model/SwatchAttributesProvider.php [hunk 12]
```diff
@@ -86,4 +96,16 @@ public function provide(Product $product)
...
     }
+
+    /**
+     * This method introduced only for the case when customer already has converted attribute.
+     *
+     * @param $productAttribute
+     * @return bool
+     * @deprecated
+     */
+    private function hasSwatchAttribute($productAttribute)
```
Comment (sidolov) [inline app/code/Magento/Swatches/Model/SwatchAttributesProvider.php]: I think this method left from the previous implementation and useless for now, please, revert

File: app/code/Magento/Swatches/Test/Unit/Model/SwatchAttributesProviderTest.php [hunk 13]
```diff
@@ -44,14 +49,21 @@ protected function setUp()
...
-            ['getConfigurableAttributes', 'getCodes']
+            ['getConfigurableAttributes', 'getCodes', 'getProductAttribute']
         );
 
         $this->swatchAttributeCodes = $this->createMock(SwatchAttributeCodes::class);
 
         $this->productMock = $this->createPartialMock(\Magento\Catalog\Model\Product::class, ['getId', 'getTypeId']);
         $this->swatchTypeChecker = $this->createMock(SwatchAttributeType::class);
 
+//        $this->swatchHelper = (new ObjectManager($this))
```
Comment (sidolov) [inline app/code/Magento/Swatches/Test/Unit/Model/SwatchAttributesProviderTest.php]: Please, remove commented code

**Review by sidolov — CHANGES_REQUESTED** [review]

**Comment by sidolov** [comment]
Hi @eugene-shab , just a few fixes needed and please, take a look on failed tests

**Comment by eugene-shab** [comment]
Hi @sidolov - fixed.

File: app/code/Magento/Swatches/Model/SwatchAttributesProvider.php [hunk 14]
```diff
@@ -17,6 +18,10 @@
  */
 class SwatchAttributesProvider
 {
+    /**
+     * @var \Magento\Swatches\Helper\Data
+     */
+    protected $swatchesHelper;
```
Comment (sidolov) [inline app/code/Magento/Swatches/Model/SwatchAttributesProvider.php]: Looks like this dependency is redundant, please, remove it, and remove usage as constructor argument

File: app/code/Magento/Swatches/etc/di.xml [hunk 15]
```diff
@@ -81,4 +81,9 @@
             </argument>
         </arguments>
     </type>
+    <type name="\Magento\Swatches\Model\SwatchAttributesProvider">
```
Comment (sidolov) [inline app/code/Magento/Swatches/etc/di.xml]: According to the previous comment seems like this configuration is useless

File: app/code/Magento/Swatches/Test/Unit/Model/SwatchAttributesProviderTest.php [hunk 16]
```diff
@@ -30,6 +30,11 @@ class SwatchAttributesProviderTest extends \PHPUnit\Framework\TestCase
      */
     private $swatchAttributeCodes;
 
+    /**
+     * @var \Magento\Swatches\Helper\Data|\PHPUnit_Framework_MockObject_MockObject
+     */
+    private $swatchHelper;
```
Comment (sidolov) [inline app/code/Magento/Swatches/Test/Unit/Model/SwatchAttributesProviderTest.php]: This changes not actual according to the previous comment

**Review by sidolov — CHANGES_REQUESTED** [review]

**Comment by sidolov** [comment]
@eugene-shab , looks good! Please, revert a couple of changes from previous implementation

**Comment by eugene-shab** [comment]
Great thanks @sidolov for code reviewing. I updated the code regarding your comments, please check.

**Review by sidolov — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sidolov, thank you for the review.
[ENGCOM-1443](https://jira.corp.magento.com/browse/ENGCOM-1443) has been created to process this Pull Request

**Comment by ishakhsuvarov** [comment]
@magento-engcom-team give me 2.2-develop instance

**Comment by magento-engcom-team** [comment]
Hi @ishakhsuvarov. Thank you for your request. I'm working on Magento 2.2-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @ishakhsuvarov, here is your [Magento instance](https://i-12771-2-2-develop.engcom.dev.magento.com).
Admin access: [https://i-12771-2-2-develop.engcom.dev.magento.com/admin](https://i-12771-2-2-develop.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by ishakhsuvarov** [comment]
@magento-engcom-team give me new test instance

**Comment by magento-engcom-team** [comment]
Hi @ishakhsuvarov. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @ishakhsuvarov, here is your [new Magento instance](https://pr-12771.engcom.dev.magento.com).
Admin access: [https://pr-12771.engcom.dev.magento.com/admin](https://pr-12771.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`
Please make sure you are PR **author** or **assignee** to access the instance.

**Comment by magento-engcom-team** [comment]
Hi @eugene-shab. Thank you for your contribution. 
Changes from your Pull Request will be available with the upcoming 2.2.6 release.

Please, consider to port this solution to 2.3 release line.
You may use [Porting tool](https://github.com/magento/magento2/wiki/Simple-way-to-port-your-solutions-automatically) to port commits automatically.

**Pull request merged**